hostpolicy/api/permissions.py

```diff
@@ -13,15 +16,58 @@ def is_super_or_hostpolicy_admin(user):
 
 class IsSuperOrHostPolicyAdminOrReadOnly(IsAuthenticated):
     """
-    Permit user if in super or group admin group, else read only.
+    Permit user if in super or group admin group, or has been granted access through a
+    NetGroupRegexPermission, else read only.
     """
 
     def has_permission(self, request, view):
         if not super().has_permission(request, view):
+            # Not even reading is allowed if you're not authenticated
             return False
         if request.method in SAFE_METHODS:
             return True
-        return is_super_or_hostpolicy_admin(request.user)
+        if is_super_or_hostpolicy_admin(request.user):
+            return True
+
+        # Is this request about atoms or something else that isn't a role?
+        # In that case, non-admin-users shouldn't have access anyway, and we can deny the request.
+        if not (view.__class__.__name__ == 'HostPolicyRoleHostsDetail' or
+                view.__class__.__name__ == 'HostPolicyRoleHostsList'):
+            return False
+
+        # Find out which labels are attached to this role
+        role = HostPolicyRole.objects.get(name=view.kwargs['name'])
+        role_labels = []
+        for label in role.labels.all():
+            role_labels.append(label.name)
+
+        # Find all the NetGroupRegexPermission objects that correspond with
+        # the ipaddress, hostname, and the groups that the user is a member of
+        if 'host' in view.kwargs:
+            hostname = view.kwargs['host']
+        else:
+            hostname = request.data.get("name")
+        host = Host.objects.get(name=hostname)
+        ips = []
+        hostserr = HostSerializer(host)
+        for i in hostserr.data['ipaddresses']:
+            ips.append(i['ipaddress'])
+        qs = NetGroupRegexPermission.find_perm(request.user.group_list, host.name, ips)
+
+        # If no permissions matched the host/ip, we deny access
+        if not qs.exists():
+            return False
+
+        # Do any of those permissions have labels that match the labels attached to this role?
+        # If so, access is granted
+        for perm in qs:
+            for label in perm.labels.all():
+                if label.name in role_labels:
+                    return True
+
+        # If the code got to this point, it means none of the labels matched.
+        return False
+
 
     def has_m2m_change_permission(self, request, view):
-        return is_super_or_hostpolicy_admin(request.user)
+        return True
```

Comment:
Seems to permissive, but might be fine?

Reply:
Pretty sure this OK. It is just an early check to see if the request can rejected before doing the heavy lifting.

Reply:
I'm not familiar with these parts. Couldn't the function call just be removed? Why have a permission that always return True?

Reply:
Could have been if there was a parent class which provided these by default, but this class just inherits `IsAuthenticated` and all the `m2m`-stuff is mreg-only code.
The usage is https://github.com/unioslo/mreg/blob/feature-hostpolicy-labels/mreg/api/v1/views_m2m.py#L14
It is modeled similar to Django's non-m2m objects.